Add charm architecture document (#89)
canonical/gateway-api-integrator-operator: docs/explanation/charm-architecture.md

# Charm architecture

> The `gateway-api-integrator` is very similar to the [`gateway-api-integrator`](https://charmhub.io/gateway-api-integrator)

Reviewer: This doesn't make sense to me -- I guess because of the two mentions of `gateway-api-integrator`. Are you referring to two different `gateway-api-integrator` charms here? I tried to visit https://charmhub.io/gateway-api-integrator-operator but that link seems to be broken.

charm in that it provides Kubernetes ingress to charms that require it. However,
instead of interacting with the Kubernetes cluster using the Ingress API, it 
uses the newer [Gateway API](https://kubernetes.io/docs/concepts/services-networking/gateway/)
, which offers a more portable, expressive, and extensible integration with all
types of Kubernetes clusters.

The Gateway API integrator receives ingress requests using the [`ingress` relation](https://github.com/canonical/charm-relation-interfaces/blob/main/interfaces/ingress/v2/README.md).  
It relies on the `tls-certificates` relation to gather TLS certificates for 
the gateway's TLS configuration. Information about the backend application is 
collected from the `ingress` relation and combined with the TLS certificates 

> from the `tls-certificates` relation. As a result, Kubernetes [`Service` resources](https://kubernetes.io/docs/concepts/services-networking/service/),  
> [`Gateway` resources](https://gateway-api.sigs.k8s.io/concepts/api-overview/#gateway),  

Reviewer: ```suggestion
from the `tls-certificates` relation. As a result, Kubernetes [`Service` resources](https://kubernetes.io/docs/concepts/services-networking/service/),
[`Gateway` resources](https://gateway-api.sigs.k8s.io/concepts/api-overview/#gateway),
```
With those empty spaces at the end of these lines, the rendered version of this documentation places "`Gateway` resources" and "`HTTPRoute` resources" on new lines

[`HTTPRoute` resources](https://gateway-api.sigs.k8s.io/concepts/api-overview/#httproute),
and [`Secret` resources](https://kubernetes.io/docs/concepts/configuration/secret/)  
will be created to facilitate ingress as expected by the ingress requirer.

In designing this charm, we've leveraged the [sidecar](https://kubernetes.io/blog/2015/06/the-distributed-system-toolkit-patterns/#example-1-sidecar-containers) pattern for Kubernetes 
charms, but somewhat unusually we're not actually deploying a workload container
alongside our charm code. Instead, the charm code talks directly to the 
Kubernetes API to provision the appropriate gateway API resource to enable 
traffic to reach the service in question. 

> As a result, if you run a `kubectl get pods` on a namespace named for the Juju 

Reviewer: ```suggestion
As a result, if you run `kubectl get pods` on a namespace named for the Juju 
```

model you’ve deployed the gateway-api-integrator charm into, you’ll see 
something like the following:

```
NAME                             READY   STATUS    RESTARTS   AGE
gateway-api-integrator-0       1/1     Running   0          3h47m

> ```

Reviewer: Nitpick: it may be nice to add a sentence after this code block explaining what this block is. Something like, `This shows there is only one container for the charm code.`

7. [`certificate-invalidated` from `tls_certificates` charm library](https://charmhub.io/tls-certificates-interface/libraries/tls_certificates)
8. [`all-certificates-invalidated` from `tls_certificates` charm library](https://charmhub.io/tls-certificates-interface/libraries/tls_certificates)
9. [`get-certificate-action`](https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/hook/#action-action)
10. [`certificates-relation-created`](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#endpoint-relation-created)
11. [`certificates-relation-joined`](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#endpoint-relation-joined)
12. [`certificates-relation-broken`](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#endpoint-relation-broken)

## Charm code overview

The `src/charm.py` is the default entry point for a charm and has the 
`GatewayAPICharm` Python class which inherits from CharmBase. CharmBase is the 
base class from which all Charms are formed, defined by [Ops](https://juju.is/docs/sdk/ops)
(Python framework for developing charms).

> > See more in the Juju docs: [Charm](https://juju.is/docs/sdk/constructs#heading--charm)

Reviewer: ```suggestion
> See more in the Juju docs: [Charm](https://documentation.ubuntu.com/juju/3.6/reference/charm/)
```

            Component(tls-certificates-lib, "tls_certificates_interface.v3.tls_certificates")
            Component(ingress-lib, "traefik_k8s.v2.ingress")
        }
        
        Container_Boundary(charm, "Charm Logic") {
            Component(gateway-api-integrator-charm, "Gateway API Integrator Charm")
        }
        
        Container_Boundary(kubernetes-lib, "Kubernetes Libraries") {
            Component(gateway, "Gateway")
            Component(http-route, "HTTPRoute")
            Component(secret, "Secret")
            Component(service, "Service")
        }
        
        Rel(tls-certificates-lib, gateway-api-integrator-charm, "Server Cert")
        Rel(ingress-lib, gateway-api-integrator-charm, "Ingress Spec")
        Rel(gateway-api-integrator-charm, gateway, "Reconciliation")
        Rel(gateway-api-integrator-charm, http-route, "Reconciliation")
        Rel(gateway-api-integrator-charm, secret, "Reconciliation")
        Rel(gateway-api-integrator-charm, service, "Reconciliation")

        UpdateLayoutConfig($c4ShapeInRow="1", $c4BoundaryInRow="4")
        
        UpdateRelStyle(tls-certificates-lib, gateway-api-integrator-charm, $offsetY="10", $offsetX="-40")

>         UpdateRelStyle(ingress-lib, gateway-api-integrator-charm, $offsetY="0", $offsetX="-40")

Reviewer: ```suggestion
        UpdateRelStyle(ingress-lib, gateway-api-integrator-charm, $offsetY="20", $offsetX="-80")
```
Nitpick

# Charm architecture

> The `gateway-api-integrator` is very similar to the [`nginx-ingress-integrator`](https://charmhub.io/nginx-ingress-integrator)

Reviewer: ```suggestion
The `gateway-api-integrator` charm is very similar to the [`nginx-ingress-integrator`](https://charmhub.io/nginx-ingress-integrator)
```

charm in that it provides Kubernetes ingress to charms that require it. However,
instead of interacting with the Kubernetes cluster using the Ingress API, it 
uses the newer [Gateway API](https://kubernetes.io/docs/concepts/services-networking/gateway/)
, which offers a more portable, expressive, and extensible integration with all
types of Kubernetes clusters.

The Gateway API integrator receives ingress requests using the [`ingress` relation](https://github.com/canonical/charm-relation-interfaces/blob/main/interfaces/ingress/v2/README.md).  
It relies on the `tls-certificates` relation to gather TLS certificates for 
the gateway's TLS configuration. Information about the backend application is 
collected from the `ingress` relation and combined with the TLS certificates 
from the `tls-certificates` relation. As a result, Kubernetes [`Service` resources](https://kubernetes.io/docs/concepts/services-networking/service/),
[`Gateway` resources](https://gateway-api.sigs.k8s.io/concepts/api-overview/#gateway),
[`HTTPRoute` resources](https://gateway-api.sigs.k8s.io/concepts/api-overview/#httproute),

> and [`Secret` resources](https://kubernetes.io/docs/concepts/configuration/secret/)  

Reviewer: ```suggestion
and [`Secret` resources](https://kubernetes.io/docs/concepts/configuration/secret/)
```
Nitpick; it's not that obvious, but the trailing whitespace at the end of this line forces the `will be created...` text to a new line in the rendered output.

9. [`get-certificate-action`](https://canonical-juju.readthedocs-hosted.com/en/latest/user/reference/hook/#action-action)
10. [`certificates-relation-created`](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#endpoint-relation-created)
11. [`certificates-relation-joined`](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#endpoint-relation-joined)
12. [`certificates-relation-broken`](https://documentation.ubuntu.com/juju/latest/reference/hook/index.html#endpoint-relation-broken)

## Charm code overview

The `src/charm.py` is the default entry point for a charm and has the 
`GatewayAPICharm` Python class which inherits from CharmBase. CharmBase is the 
base class from which all Charms are formed, defined by [Ops](https://juju.is/docs/sdk/ops)
(Python framework for developing charms).

> See more in the Juju docs: [Charm](https://documentation.ubuntu.com/juju/3.6/reference/charm/)

Take, for example, when a configuration is changed by using the CLI.

1. User runs the configuration command:
```bash
juju config <relevant-charm-configuration>

> ```

Reviewer: Could you fill in a relevant configuration example for gateway-api-integrator?

Author: Yes, no problem, added, thanks!
